libraries/AP_Torqeedo/AP_Torqeedo_TQBus.cpp

```diff
@@ -0,0 +1,1119 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "AP_Torqeedo_TQBus.h"
+
+#if AP_TORQEEDO_TQBUS_ENABLED
+#include <AP_Common/AP_Common.h>
+#include <AP_Math/AP_Math.h>
+#include <SRV_Channel/SRV_Channel.h>
+#include <AP_Logger/AP_Logger.h>
+#include <GCS_MAVLink/GCS.h>
+#include <stdio.h>
+
+#define TORQEEDO_SERIAL_BAUD        19200   // communication is always at 19200
+#define TORQEEDO_PACKET_HEADER      0xAC    // communication packet header
+#define TORQEEDO_PACKET_FOOTER      0xAD    // communication packet footer
+#define TORQEEDO_PACKET_ESCAPE      0xAE    // escape character for handling occurrences of header, footer and this escape bytes in original message
+#define TORQEEDO_PACKET_ESCAPE_MASK 0x80    // byte after ESCAPE character should be XOR'd with this value
+#define TORQEEDO_LOG_TRQD_INTERVAL_MS                   5000// log TRQD message at this interval in milliseconds
+#define TORQEEDO_SEND_MOTOR_SPEED_INTERVAL_MS           100 // motor speed sent at 10hz if connected to motor
+#define TORQEEDO_SEND_MOTOR_STATUS_REQUEST_INTERVAL_MS  400 // motor status requested every 0.4sec if connected to motor
+#define TORQEEDO_SEND_MOTOR_PARAM_REQUEST_INTERVAL_MS   400 // motor param requested every 0.4sec if connected to motor
+#define TORQEEDO_BATT_TIMEOUT_MS    5000    // battery info timeouts after 5 seconds
+#define TORQEEDO_REPLY_TIMEOUT_MS   25      // stop waiting for replies after 25ms
+#define TORQEEDO_ERROR_REPORT_INTERVAL_MAX_MS   10000   // errors reported to user at no less than once every 10 seconds
+
+extern const AP_HAL::HAL& hal;
+
+// parameters
+const AP_Param::GroupInfo AP_Torqeedo_TQBus::var_info[] = {
+
+    // @Param: TYPE
+    // @DisplayName: Torqeedo connection type
+    // @Description: Torqeedo connection type
+    // @Values: 0:Disabled, 1:Tiller, 2:Motor
+    // @User: Standard
+    // @RebootRequired: True
+    AP_GROUPINFO_FLAGS("TYPE", 1, AP_Torqeedo_TQBus, _type, (int8_t)ConnectionType::TYPE_DISABLED, AP_PARAM_FLAG_ENABLE),
```

Comment:
Is it correct that we have a TYPE parameter defined both here and in AP_Torqeedo_Params.h/.cpp?